lib/interface_quda.cpp

```diff
@@ -4146,6 +4154,74 @@ int computeGaugePathQuda(void *out, void *siteLink, int ***input_path_buf, int *
   return 0;
 }
 
+int computeGaugeLoopTraceQuda(double _Complex* traces, void *siteLink, int **input_path_buf, int *path_length, double *loop_coeff,
+                                    int num_paths, int max_length, double factor, QudaGaugeParam *qudaGaugeParam)
+{
+  profileGaugeLoopTrace.TPSTART(QUDA_PROFILE_TOTAL);
+  profileGaugeLoopTrace.TPSTART(QUDA_PROFILE_INIT);
+
+  checkGaugeParam(qudaGaugeParam);
+
+  GaugeFieldParam gParam(*qudaGaugeParam, siteLink);
+  gParam.location = QUDA_CPU_FIELD_LOCATION;
+  gParam.site_offset = qudaGaugeParam->gauge_offset;
+  gParam.site_size = qudaGaugeParam->site_size;
+  cpuGaugeField *cpuSiteLink = (!qudaGaugeParam->use_resident_gauge) ? new cpuGaugeField(gParam) : nullptr;
+
+  cudaGaugeField *cudaSiteLink = nullptr;
+
+  if (qudaGaugeParam->use_resident_gauge) {
+    if (!gaugePrecise) errorQuda("No resident gauge field to use");
+    cudaSiteLink = gaugePrecise;
+  } else {
+    gParam.location = QUDA_CUDA_FIELD_LOCATION;
+    gParam.create = QUDA_NULL_FIELD_CREATE;
+    gParam.reconstruct = qudaGaugeParam->reconstruct;
+    gParam.setPrecision(qudaGaugeParam->cuda_prec, true);
+
+    cudaSiteLink = new cudaGaugeField(gParam);
+    profileGaugeLoopTrace.TPSTOP(QUDA_PROFILE_INIT);
+
+    profileGaugeLoopTrace.TPSTART(QUDA_PROFILE_H2D);
+    cudaSiteLink->loadCPUField(*cpuSiteLink);
+    profileGaugeLoopTrace.TPSTOP(QUDA_PROFILE_H2D);
+
+    profileGaugeLoopTrace.TPSTART(QUDA_PROFILE_INIT);
+  }
+  profileGaugeLoopTrace.TPSTOP(QUDA_PROFILE_INIT);
+
+  cudaGaugeField *cudaGauge = createExtendedGauge(*cudaSiteLink, R, profileGaugeLoopTrace);
+  // apply / remove phase as appropriate
+  if (cudaGauge->StaggeredPhaseApplied()) cudaGauge->removeStaggeredPhase();
+
+  // prepare trace storage
+  std::vector<Complex> loop_traces(num_paths);
+
+  // actually do the computation
+  profileGaugeLoopTrace.TPSTART(QUDA_PROFILE_COMPUTE);
+  gaugeLoopTrace(*cudaGauge, loop_traces, factor, &input_path_buf, path_length, loop_coeff, num_paths, max_length);
+  profileGaugeLoopTrace.TPSTOP(QUDA_PROFILE_COMPUTE);
+
+  for (int i = 0; i < num_paths; i++) { memcpy(traces + i, &loop_traces[i], sizeof(Complex)); }
+
+  profileGaugeLoopTrace.TPSTART(QUDA_PROFILE_FREE);
+  if (qudaGaugeParam->make_resident_gauge) {
+    if (gaugePrecise && gaugePrecise != cudaSiteLink) delete gaugePrecise;
+    gaugePrecise = cudaSiteLink;
+    if (extendedGaugeResident) delete extendedGaugeResident;
+    extendedGaugeResident = cudaGauge;
+  } else {
+    delete cudaSiteLink;
+    delete cudaGauge;
+  }
+
+  if (cpuSiteLink) delete cpuSiteLink;
+  profileGaugeLoopTrace.TPSTOP(QUDA_PROFILE_FREE);
+
+  profileGaugeLoopTrace.TPSTOP(QUDA_PROFILE_TOTAL);
+  return 0;
```

Comment:
What's the point of the `return 0`?

Reply:
It was just a copy + paste of what's done in the gauge path routine, but it isn't necessary there either, so I can poke that out in a separate commit.

Reply:
https://github.com/lattice/quda/pull/1308/commits/04f84f575b3a13a6d8e1bf0c520553f4c038c033